src/detect-dnp3.c

```diff
@@ -443,11 +438,10 @@ static int DetectDNP3ObjMatch(DetectEngineThreadCtx *det_ctx,
     DetectDNP3 *detect = (DetectDNP3 *)ctx;
     DNP3ObjectList *objects = NULL;
 
-    if (flags & STREAM_TOSERVER) {
-        objects = &tx->request_objects;
-    }
-    else if (flags & STREAM_TOCLIENT) {
-        objects = &tx->response_objects;
+    if (flags & STREAM_TOSERVER && tx->is_request) {
+        objects = &tx->objects;
+    } else if (flags & STREAM_TOCLIENT && !tx->is_request) {
+        objects = &tx->objects;
     }
```

Comment:
Its code like this that fails when `APP_LAYER_PARSER_OPT_UNIDIR_TXS` is set for the protocol.  When set the direction logic appears to be inversed. The detect function is never called with `STREAM_TOSERVER` for a request transaction, and vice-versa, as these detect functions are being called the TCP ACK.  Running in IPS mode makes the direction logic correct, which I'm still trying to figure out.  I believe the progress logic is correct at https://github.com/OISF/suricata/pull/8006/files#diff-e6b6ada9b3a17879876cc1c3a1db5f162e3f834cc731bf3eab4d02bde39ae424R1437

However, the parser, logging and detect seems to work find without setting the `APP_LAYER_PARSER_OPT_UNIDIR_TXS`, so it might not be needed.